modAionImpl/src/org/aion/zero/impl/sync/msg/ResponseTrieData.java

```diff
@@ -0,0 +1,259 @@
+    private final ByteArrayWrapper nodeKey; // data of 32 bytes
+    private final byte[] nodeValue;
+    private final Map<ByteArrayWrapper, byte[]> referencedNodes; // empty for leaf nodes
+    private final DatabaseType dbType;
+
+    /**
+     * Constructor for trie node responses that represent leafs, i.e. a single key-value pair.
+     *
+     * @param nodeKey the key of the requested trie node
+     * @param nodeValue the value stored for the requested trie node
+     * @param dbType the blockchain database in which the key should be found
+     * @throws NullPointerException if any of the given parameters are {@code null}
+     */
+    public ResponseTrieData(
+            final ByteArrayWrapper nodeKey, final byte[] nodeValue, final DatabaseType dbType) {
+        super(Ver.V1, Ctrl.SYNC, Act.RESPONSE_TRIE_DATA);
+
+        // ensure inputs are not null
+        Objects.requireNonNull(nodeKey);
+        Objects.requireNonNull(nodeValue);
+        Objects.requireNonNull(dbType);
+
+        this.nodeKey = nodeKey;
+        this.nodeValue = nodeValue;
+        this.referencedNodes = new HashMap<>();
+        this.dbType = dbType;
+    }
+
+    /**
+     * Constructor for trie node responses that contain multiple nodes (key-value pairs).
+     *
+     * @param nodeKey the key of the requested trie node
+     * @param nodeValue the value stored for the requested trie node
+     * @param referencedNodes a map of key-value pairs referenced by the value of the requested key
+     * @param dbType the blockchain database in which the key should be found
+     * @throws NullPointerException if any of the given parameters are {@code null}
+     * @implNote The referenced nodes are purposefully not deep copied to minimize resource
+     *     instantiation. This is a reasonable choice given that these objects are created to be
+     *     encoded and transmitted over the network and therefore there is the expectation that they
+     *     will not be utilized further.
+     */
+    public ResponseTrieData(
+            final ByteArrayWrapper nodeKey,
+            final byte[] nodeValue,
+            final Map<ByteArrayWrapper, byte[]> referencedNodes,
+            final DatabaseType dbType) {
+        super(Ver.V1, Ctrl.SYNC, Act.RESPONSE_TRIE_DATA);
+
+        // ensure inputs are not null
+        Objects.requireNonNull(nodeKey);
+        Objects.requireNonNull(nodeValue);
+        Objects.requireNonNull(dbType);
+        Objects.requireNonNull(referencedNodes);
+
+        this.nodeKey = nodeKey;
+        this.nodeValue = nodeValue;
+        this.referencedNodes = referencedNodes;
+        this.dbType = dbType;
+    }
+
+    /**
+     * Decodes a message into a trie node response.
+     *
+     * @param message a {@code byte} array representing a response to a trie node request.
+     * @return the decoded trie node response
+     * @implNote The decoder must return {@code null} if any of the component values are {@code
```

Comment:
Would it make sense to log something in case of component values being invalid?

Seems like otherwise it'd be really hard to debug if something in the kernel set up the input wrong earlier in the call stack

Reply:
The unit tests cover various correct/incorrect scenarios for building the object. They would fail if the implementation changes to violate the expected behavior.

Reply:
I meant if someone who is writing new code that calls ResponseTrieData.decode( .. ) -- they can still use it wrong and it wouldn't be covered by any existing test.  I don't debug this part of the code much, so I'll defer to your judgement ... it just seems like for someone who is implementing code that calls this function, or debugging a log file that touches this, having log messages for those unexpected cases would make debugging a lot easier if things go wrong.  I don't think there's any drawback to having logs at say ... debug or trace level, if polluting the output is a concern.

Reply:
I don't think the place to log this information is inside the message, but I will expand the logged information done by the handlers in the next two PRs.